proposals/2022-09-14-kubernetes-versioning.md

```diff
@@ -0,0 +1,237 @@
+        name: wasmtime-slight-v1
+      handler: slight-v1-2-3
+      scheduling:
+        nodeSelector:
+          wasmtime-slight-v1: "true"
+      ---
+      # This runtime class maps "plugins.cri.containerd.runtimes.slight-v1" to
+      # runtime_type = "io.containerd.slight-v1-2-3.v1".
+      # 
+      # This node selector is intended select any node that has the slight shim enabled
+      # regardless of the version of the shim. As new versions are introduced, this
+      # node selector would allow for new versions of the shim to be introduced without
+      # having to update the pod workload to target a new shim version.
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-slight
+      handler: slight-v1
+      scheduling:
+        nodeSelector:
+          wasmtime-slight-enabled: "true"
+      ```
+
+    **NOTE: node labels would be expected for both `wasmtime-slight-v1: "true"` and `wasmtime-slight-v1-2-3: "true"` on nodes that have the slight shim v1.2.3 installed.**
+
+4. A Wasm workload with the specified RuntimeClass name matching the "wasmtime-slight-v1" runtime class from step 3. The "wasmtime-slight-v1" runtimeClassName would map to the slight v1.2.3 handler.
+      ```yaml
+      apiVersion: apps/v1
+      kind: Deployment
+      metadata:
+        name: wasm-slight
+      spec:
+        replicas: 1
+        selector:
+          matchLabels:
+            app: wasm-slight
+        template:
+          metadata:
+            labels:
+              app: wasm-slight
+          spec:
+            runtimeClassName: wasmtime-slight-v1
+            containers:
+            - name: slight-hello
+              image: ghcr.io/deislabs/containerd-wasm-shims/examples/slight-rust-hello:latest
+              command: ["/"]
+      ```
+
+### <a name='Introducinganewversionofashim'></a>Introducing a new version of a shim
+Now that we have described the versioning scheme, let's describe what happens when we introduce a new node that contains a new version of a shim. In this example, we will introduce v1.2.4 of the slight shim.
+
+For this example, let's imagine we have a cluster with 1 node configured as described in [Versioning of both shim binary and RuntimeClass name](#versioning-of-both-shim-binary-and-runtimeclass-name). We introduce a new node to the cluster with `/bin/containerd-shim-slight-v1-2-4-v1` installed and the following containerd config section.
+
+```toml
+[plugins.cri.containerd.runtimes.slight-v1-2-4]
+    runtime_type = "io.containerd.slight-v1-2-4.v1" 
+```
+
+This new node running v1.2.4 would be decorated with the following node labels, `wasmtime-slight-v1: true`, `wasmtime-slight-enabled: true`, and `wasmtime-slight-v1.2.4: true`. 
+
+**NOTE:** `wasmtime-slight-v1` node label is shared between the node running v1.2.3 and the node running v1.2.4. This means the RuntimeClasses of the same names will match scheduling labels on both the new node and the old node.
```

Comment:
How do we make sure forward compatibility within the same major version here? 

Let's say a node is running `v.1.2` shim and the image is expecting `v1.3` slight shim which has new features that `v.1.2` doesn't have.

Reply:
I was trying to simplify things, but I think you are right. We must provide the ability to say the minor version to provide feature selectivity to workloads.

I've added a major+minor label and related content.